agent/registration/csr.go

```diff
@@ -4,93 +4,133 @@
-		if !apierrors.IsNotFound(err) {
-			klog.Errorf("error getting csr %s, err=%v", hostname, err)
-			return nil, err
-		}
-		privKey, byoCSR, err = bcsr.generateCSR(hostname)
-		if err != nil {
-			klog.Errorf("error generating csr %s, err=%v", hostname, err)
-			return nil, err
-		}
-		err = bcsr.K8sClient.Create(ctx, byoCSR)
-		if err != nil {
-			klog.Errorf("error creating host csr %s, err=%v", hostname, err)
-			return nil, err
-		}
+func (bcsr *ByohCSR) RequestBYOHClientCert(hostname string) (string, types.UID, error) {
+	if hostname == "" {
+		return "", "", fmt.Errorf("hostname is not valid")
 	}
-	return privKey, nil
-}
-
-func (bcsr *ByohCSR) generateCSR(hostname string) (*rsa.PrivateKey, *certv1.CertificateSigningRequest, error) {
 	// Generate Private Key
 	privateKey, err := rsa.GenerateKey(rand.Reader, KeySize)
 	if err != nil {
-		return nil, nil, err
+		return "", "", err
 	}
+	bcsr.PrivateKey = privateKey
+	csrData, err := generateCSR(hostname, privateKey)
+	if err != nil {
+		klog.Errorf("error generating csr %s, err=%v", hostname, err)
+		return "", "", err
+	}
+	certTimeToExpire := time.Duration(ExpirationSeconds) * time.Second
+	reqName, reqUID, err := csr.RequestCertificate(bcsr.BootstrapClient,
+		csrData,
+		fmt.Sprintf(ByohCSRNameFormat, hostname),
+		certv1.KubeAPIServerClientSignerName,
+		&certTimeToExpire,
+		[]certv1.KeyUsage{certv1.UsageClientAuth},
+		privateKey)
+	if err != nil {
+		return "", "", err
+	}
+	return reqName, reqUID, nil
+}
 
+func generateCSR(hostname string, privKey *rsa.PrivateKey) ([]byte, error) {
 	// Generate a new *x509.CertificateRequest template
-	// TODO: validate template
 	csrTemplate := x509.CertificateRequest{
 		Subject: pkix.Name{
 			CommonName:   fmt.Sprintf(ByohCSRCNFormat, hostname),
-			Organization: []string{"byoh:hosts"},
+			Organization: []string{ByohCSROrg},
 		},
 	}
-
 	// Generate the CSR bytes
-	csrData, err := x509.CreateCertificateRequest(rand.Reader, &csrTemplate, privateKey)
+	csrData, err := x509.CreateCertificateRequest(rand.Reader, &csrTemplate, privKey)
 	if err != nil {
-		return nil, nil, err
+		return nil, err
 	}
-	// Create the CSR object
-	csr := &certv1.CertificateSigningRequest{
-		ObjectMeta: metav1.ObjectMeta{
-			Name:        fmt.Sprintf(ByohCSRNameFormat, hostname),
-			Labels:      map[string]string{},
-			Annotations: map[string]string{},
-		},
-		Spec: certv1.CertificateSigningRequestSpec{
-			Request:           pem.EncodeToMemory(&pem.Block{Type: "CERTIFICATE REQUEST", Bytes: csrData}),
-			SignerName:        certv1.KubeAPIServerClientSignerName,
-			Usages:            []certv1.KeyUsage{certv1.UsageClientAuth},
-			ExpirationSeconds: pointer.Int32(ExpirationSeconds),
-		},
+	csrPemBlock := &pem.Block{
+		Type:  cert.CertificateRequestBlockType,
+		Bytes: csrData,
 	}
-	return privateKey, csr, nil
+	return pem.EncodeToMemory(csrPemBlock), nil
+}
+
+func LoadRESTClientConfig(bootstrapKubeconfig string) (*restclient.Config, error) {
```

Comment:
Would be good to add a small description for public functions.

Reply:
fixed.